### PR TITLE
RM-9025 Support ARIA 1.2 for the BocAutoCompleteReferenceValue

### PR DESCRIPTION
**Pull request by a-ctor**
https://re-motion.atlassian.net/browse/RM-9025

Solution implemented with the reference implementation https://raw.githack.com/w3c/aria-practices/aria1.2-combobox-proposal/examples/combobox/combobox-autocomplete-list.html in mind. Tested with JAWS 2021, JAWS 2023, NVDA, and Narrator. See repro in the issue description.

Open points @MichaelKetting:
 - [x] "Update WCAG samples/doc for combobox behavior." What samples/doc?

### CODE REVIEW

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts [hunk 1]
```diff
@@ -234,11 +235,7 @@ namespace Remotion.BocAutoCompleteReferenceValue
             serviceMethodSearch: serviceMethodSearch,
             serviceMethodSearchExact: serviceMethodSearchExact,
             data: null,
-            combobox: null,
-            selectListID: null,
-            informationPopUpID: null,
             // re-motion: clicking this control will display the dropdown list with an assumed input of '' (regardless of textbox value)
-            dropDownButtonID: null
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts:241]: Why did you remove those?
And if removeing it is sensible, why leave the comment?

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts [hunk 2]
```diff
@@ -1480,13 +1472,13 @@ namespace Remotion.BocAutoCompleteReferenceValue
...
-            this.options.combobox.setAttribute("aria-expanded", "false");
-            this.input.removeAttribute("aria-activedescendant");
+            this.input.setAttribute("aria-expanded", "false");
+            this.input.setAttribute("aria-activedescendant", "");
             this.element && LayoutUtility.Hide(this.element);
             if (this.listItems) {
                 for (const listItem of this.listItems) {
                     listItem.classList.remove(this.CLASSES.ACTIVE);
-                    listItem.setAttribute("aria-selected", "false")
+                    listItem.removeAttribute("aria-selected")
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts:1481]: why do you remove it instead of setting it `false`? same at the other places where it gets removed.

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts [hunk 3]
```diff
@@ -1480,13 +1472,13 @@ namespace Remotion.BocAutoCompleteReferenceValue
         public hide() {
             if (this.repositionTimer) 
                 clearTimeout(this.repositionTimer);
-            this.options.combobox.setAttribute("aria-expanded", "false");
-            this.input.removeAttribute("aria-activedescendant");
+            this.input.setAttribute("aria-expanded", "false");
+            this.input.setAttribute("aria-activedescendant", "");
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts:1476]: need to check if removing or empty is the "correct" choice here.

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts [hunk 4]
```diff
@@ -1166,19 +1163,13 @@ namespace Remotion.BocAutoCompleteReferenceValue
...
             this.element = document.createElement("div");
-            this.element.setAttribute('role', 'listbox');
-            this.element.setAttribute('id', this.options.selectListID);
             this.element.setAttribute('class', this.options.resultsClass);
             this.element.style.position = 'fixed';
             LayoutUtility.Hide(this.element);
 
             this.input.closest('div, td, th, body')!.appendChild(this.element);
 
-            this.options.combobox.setAttribute('aria-owns', this.options.selectListID);
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts:1177]: not sure if removing the `aria-owns` is correct.

File: Remotion/ObjectBinding/Web/UI/Controls/BocReferenceValueImplementation/Rendering/BocAutoCompleteReferenceValueRenderer.cs [hunk 5]
```diff
@@ -375,7 +364,17 @@ private TextBox GetTextBox (BocRenderingContext<IBocAutoCompleteReferenceValue>
       textBox.Page = renderingContext.Control.Page!.WrappedInstance;
       textBox.ApplyStyle(renderingContext.Control.CommonStyle);
       renderingContext.Control.TextBoxStyle.ApplyStyle(textBox);
+
+      textBox.Attributes.Add(HtmlTextWriterAttribute2.Role, HtmlRoleAttributeValue.Combobox);
       textBox.Attributes.Add(HtmlTextWriterAttribute2.AriaAutoComplete, HtmlAriaAutoCompleteAttributeValue.Both);
+      textBox.Attributes.Add(HtmlTextWriterAttribute2.AriaExpanded, HtmlAriaExpandedAttributeValue.False);
+      textBox.Attributes.Add(HtmlTextWriterAttribute2.AriaHasPopup, HtmlAriaHasPopupAttributeValue.True);
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web/UI/Controls/BocReferenceValueImplementation/Rendering/BocAutoCompleteReferenceValueRenderer.cs:371]: why did you change this from a listbox to a boolean value?

File: Remotion/Web/Core/UI/HtmlAriaHasPopupAttributeValue.cs [hunk 6]
```diff
@@ -28,5 +28,8 @@ public static class HtmlAriaHasPopupAttributeValue
     public const string Listbox = "listbox";
     public const string Menu = "menu";
     public const string Tree = "tree";
+
+    public const string True = "true";
+    public const string False = "false";
```
Comment (MichaelKetting) [inline Remotion/Web/Core/UI/HtmlAriaHasPopupAttributeValue.cs:33]: I believe those are incorrect.

File: Remotion/ObjectBinding/Web/UI/Controls/BocReferenceValueImplementation/Rendering/BocAutoCompleteReferenceValueRenderer.cs [hunk 7]
```diff
@@ -375,7 +364,17 @@ private TextBox GetTextBox (BocRenderingContext<IBocAutoCompleteReferenceValue>
...
+      textBox.Attributes.Add(HtmlTextWriterAttribute2.AriaControls, GetDropDownButtonName(renderingContext));
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web/UI/Controls/BocReferenceValueImplementation/Rendering/BocAutoCompleteReferenceValueRenderer.cs:372]: this one is new. please explain why the textbox has a controls relationship with the button.

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
I got a couple of questions. some stylisc, some about the aria-1.2 standard.

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts (hunk 1 above)
Comment (a-ctor) [reply Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts:241]: Whoops forgot about the comment. I had TypeScript compilation problems because the properties are actually required but were registered as optional here. No idea why this has become a problem just now.

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts (hunk 3 above)
Comment (a-ctor) [reply Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts:1476]: Same here the reference implementation just clears the value and it is always there to begin with.

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts (hunk 4 above)
Comment (a-ctor) [reply Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts:1177]: > This is essentially returning to the 1.0 pattern with one small modification -- use aria-controls to specify the input-popup relationship instead of aria-owns.

See https://github.com/w3c/aria/wiki/Resolving-ARIA-1.1-Combobox-Issues. If I understand correctly, `aria-controls` replaces `aria-owns` completely in ARIA 1.2 implemenation

File: Remotion/ObjectBinding/Web/UI/Controls/BocReferenceValueImplementation/Rendering/BocAutoCompleteReferenceValueRenderer.cs (hunk 7 above)
Comment (a-ctor) [reply Remotion/ObjectBinding/Web/UI/Controls/BocReferenceValueImplementation/Rendering/BocAutoCompleteReferenceValueRenderer.cs:372]: That's just a mistake that is fixed by the JavaScript code :p it should reference the result set instead

Comment (MichaelKetting) [reply Remotion/ObjectBinding/Web/UI/Controls/BocReferenceValueImplementation/Rendering/BocAutoCompleteReferenceValueRenderer.cs:372]: So, I found a mistake in the review that will get fixed :) It was simply compensated for already and now should also get cleaned up. 
(just summarizing to make sure we are talking about the same here)

File: Remotion/ObjectBinding/Web/UI/Controls/BocReferenceValueImplementation/Rendering/BocAutoCompleteReferenceValueRenderer.cs (hunk 5 above)
Comment (MichaelKetting) [reply Remotion/ObjectBinding/Web/UI/Controls/BocReferenceValueImplementation/Rendering/BocAutoCompleteReferenceValueRenderer.cs:371]: yeah, no. I believe there's still the four options available and it's possible `true` was also always an option but "listbox" is more descriptive about the intent. 

> Elements with the role combobox have an implicit [aria-haspopup](https://www.w3.org/TR/wai-aria-1.2/#aria-haspopup) value of listbox. If the combobox popup element has a role other than [listbox](https://www.w3.org/TR/wai-aria-1.2/#listbox), authors MUST specify a value for [aria-haspopup](https://www.w3.org/TR/wai-aria-1.2/#aria-haspopup) that corresponds to the role of its popup.

So, it's actually redundant but specifying it makes things explicit. I like explicit. Also, maybe one day we have a grid in there or a tree. then we need the other options, too :)

Reading further:
> To provide backward compatibility with ARIA 1.0 content, user agents MUST treat an `aria-haspopup` value of `true` as equivalent to a value of `menu`.

We do not want `menu`.

**Review by MichaelKetting — COMMENTED** [review]
believe all questions have been answered now.

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts [hunk 8]
```diff
@@ -1617,8 +1609,8 @@ namespace Remotion.BocAutoCompleteReferenceValue
             LayoutUtility.Hide(this.element);
             this.input.closest('div, td, th, body')!.appendChild(this.element);
 
-            if (this.options.combobox.getAttribute('aria-labelledby') !== null) {
-                this.element.setAttribute("aria-labelledby", this.options.combobox.getAttribute('aria-labelledby')!);
+            if (this.input.getAttribute('aria-labelledby') !== null) {
+                this.element.setAttribute("aria-labelledby", this.input.getAttribute('aria-labelledby')!);
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts:1613]: wait a second, I think this bit of code can simply go away? orginally, it was because I already rendered the aria-labelledby onto the input box and then moved it up to the span. what is this code doing now? please check and either explain or remove because redundant.

**Review by MichaelKetting — APPROVED** [review]
I got one final cleanup bit which we should check please. only need to see/discuss if there's a surprise hidden in there.

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts (hunk 8 above)
Comment (a-ctor) [reply Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts:1613]: This code relates to the info-box and moves the `aria-labelledby` to the opened infobox. I do think that this is still relevant.

Comment (MichaelKetting) [reply Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.UI.ts:1613]: Then we're done.
